### PR TITLE
Fix several ResourceWarning: unclosed file

### PR DESCRIPTION
**Pull request by BoboTiG**
Fixes a lot of ResourceWarning like:
```python
pre_commit/repository.py:46: ResourceWarning: unclosed file <_io.TextIOWrapper name='~/.cache/pre-commit/repozvpv73iz/py_env-python3.6/.install_state_v1' mode='r' encoding='UTF-8'>
  return json.loads(io.open(filename).read())
```
I also made changes to use the `with` context manager everywhere possible to prevent any other leaks.

### CODE REVIEW

File: pre_commit/repository.py [hunk 1]
```diff
@@ -43,7 +43,8 @@ def _read_state(prefix, venv):
     if not os.path.exists(filename):
         return None
     else:
-        return json.loads(io.open(filename).read())
+        with io.open(filename) as fh:
```
Comment (asottile) [inline pre_commit/repository.py]: very very minor: the rest of the code conventionally uses `f` as a variable name here -- mind changing them all to use that?

Comment (BoboTiG) [reply pre_commit/repository.py]: Of course!

File: tests/conftest.py [hunk 2]
```diff
@@ -21,6 +21,22 @@
...
+    yield
+    warnings = tuple(
+        warning for warning in recwarn
+        # python2 + pypy warn:
+        # ImportWarning: Not importing directory '...' missing __init__.py
+        if not (
+            isinstance(warning.message, ImportWarning) and
+            str(warning.message).startswith('Not importing directory ') and
+            str(warning.message).endswith(' missing __init__.py')
+        )
```
Comment (asottile) [inline tests/conftest.py]: you probably don't need this filtering here (I would hope?) -- I think this was just due to testing a namespace package in the other library

Comment (BoboTiG) [reply tests/conftest.py]: Nope, it is needed because on Python 3.6 and 3.7 there are a lot of these warnings.

Comment (BoboTiG) [reply tests/conftest.py]: Ah. I just saw the check is not good on Python 6:
```python
python3.6/importlib/_bootstrap_external.py:426: ImportWarning: Not importing directory ~/python3.6/site-packages/aspy: missing __init__
```
There is no `.py`. I will change that.

Comment (asottile) [reply tests/conftest.py]: Ah lol of course, the same name space package is still in play

Comment (BoboTiG) [reply tests/conftest.py]: Tests are failing on Python 3.7. But the warnings list is not helpful. I updated to patch to have a more human message.

**Comment by BoboTiG** [comment]
So, there are still ResourceWarning. But I do not if it is related to this project or third-party one. I will improve the warning logging o add the source module and lineno. I will let you knwo when the PR is OK to review again :)

**Comment by BoboTiG** [comment]
I cannot run test on `python -m pytest -l -x -v tests/repository_test.py`:
```python
tests/repository_test.py:66: in <module>
    @pytest.mark.integration
../ven36/lib/python3.6/site-packages/_pytest/mark/structures.py:367: in __getattr__
    self._check(name)
../ven36/lib/python3.6/site-packages/_pytest/mark/structures.py:383: in _check
    raise AttributeError("%r not a registered marker" % (name,))
E   AttributeError: 'integration' not a registered marker
```
I installed `requirements-dev.txt`, do you know what am I missing?

**Comment by asottile** [comment]
Looks like pytest strict mode? The markers are left over from when the test suite was split and don't provide any benefits now. Feel free to remove those :)

(probably check your environment for PYTEST env vars, those are probably turning on strict mode)

**Comment by asottile** [comment]
whichever you want to do is fine with me :) (either this PR or new PR)

File: tests/languages/all_test.py [hunk 3]
```diff
@@ -1,20 +1,34 @@
...
 import inspect
+import sys
 
 import pytest
 
 from pre_commit.languages.all import all_languages
 from pre_commit.languages.all import languages
 
 
+if sys.version_info < (3,):
```
Comment (asottile) [inline tests/languages/all_test.py]: will probably need `# pragma: no cover`s for this -- elsewhere `six.PY2` is being used, let's use that here as well

File: tests/languages/all_test.py [hunk 4]
```diff
@@ -1,20 +1,34 @@
...
+    setargspec = functools.partial(
+        inspect.ArgSpec, varargs=None, keywords=None, defaults=None,
+    )
+    getargspec = inspect.getargspec
+else:
+    setargspec = functools.partial(
```
Comment (asottile) [inline tests/languages/all_test.py]: I'd probably call this `ArgSpec` or `make_argspec` personally

**Comment by asottile** [comment]
Thanks for the awesome PR!

**Pull request merged**